src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -874,21 +871,34 @@ void makeNotNullMembersMaybeNull()
                     }
                 }
 
-                bool needsDefaultInitialStateForMembers()
+                ImmutableArray<Symbol> getMembersNeedingDefaultInitialState()
                 {
                     if (_hasInitialState)
                     {
-                        return false;
+                        return ImmutableArray<Symbol>.Empty;
                     }
 
                     // We don't use a default initial state for value type instance constructors without `: this()` because
                     // any usages of uninitialized fields will get definite assignment errors anyway.
-                    if (!method.HasThisConstructorInitializer(out _) && (!method.ContainingType.IsValueType || method.IsStatic))
+                    bool hasConstructorInitializer = method.HasThisConstructorInitializer(out _);
+                    if (!hasConstructorInitializer && (!method.ContainingType.IsValueType || method.IsStatic))
                     {
-                        return true;
+                        return method.ContainingType.GetMembersUnordered();
+                    }
+
+                    if (method.IncludeFieldInitializersInBody())
+                    {
+                        return method.ContainingType.GetMembersUnordered();
+                    }
+
+                    if (hasConstructorInitializer)
+                    {
+                        // We want to assume that all required members were _not_ set by the chained constructor
+                        return method.ContainingType.AllRequiredMembers.SelectAsArray(
```

Comment:
Consider testing a scenario where we access a required property from the base type.

```cs
public class Base
{
    public required string Prop1 { get; set; }
    public string Prop2 { get; set; } = null!;
}

public class Derived : Base
{
    public Derived()
    {
        Prop1.ToString(); // warning
        Prop2.ToString(); // ok
    }
}
```